src/main/scala/cognite/spark/v1/SequenceRowsRelation.scala

```diff
@@ -0,0 +1,318 @@
+    Array(StructField("rowNumber", DataTypes.LongType, nullable = false)) ++ sequenceInfo.columns
+      .map(col => StructField(col.externalId, sequenceTypeToSparkType(col.valueType)))
+      .toList
+  )
+
+  private def query(
+      filter: SequenceRowFilter,
+      limit: Option[Int],
+      columns: Option[Seq[String]],
+      client: GenericClient[IO, Nothing]) =
+    sequenceId match {
+      case CogniteExternalId(externalId) =>
+        client.sequenceRows.queryByExternalId(
+          externalId,
+          filter.inclusiveStart,
+          filter.exclusiveEnd,
+          limit,
+          columns)
+      case CogniteInternalId(id) =>
+        client.sequenceRows.queryById(id, filter.inclusiveStart, filter.exclusiveEnd, limit, columns)
+    }
+
+  private def insert(columns: Seq[String], rows: Seq[SequenceRow], client: GenericClient[IO, Nothing]) =
+    sequenceId match {
+      case CogniteExternalId(externalId) =>
+        client.sequenceRows.insertByExternalId(externalId, columns, rows)
+      case CogniteInternalId(id) =>
+        client.sequenceRows.insertById(id, columns, rows)
+    }
+
+  def getStreams(filters: Seq[SequenceRowFilter], expectedColumns: Array[String])(
+      client: GenericClient[IO, Nothing],
+      limit: Option[Int],
+      numPartitions: Int): Seq[Stream[IO, ProjectedSequenceRow]] =
+    filters.toVector.map { filter =>
+      val requestedColumns =
+        if (expectedColumns.isEmpty) {
+          // when no columns are needed, the API does not like it, so we have to request something
+          // prefer non-string columns since they can't contain too much data
+          Array(
+            sequenceInfo.columns
+              .find(_.valueType != "STRING")
+              .getOrElse(sequenceInfo.columns.head)
+              .externalId
+          )
+        } else {
+          expectedColumns
+        }
+      val projectedRows =
+        query(filter, limit, Some(requestedColumns), client)
+          .map {
+            case (_, rows) =>
+              rows.map { r =>
+                val values = expectedColumns
+                  .zip(r.values)
+                  .map {
+                    case (column, value) =>
+                      parseJsonValue(value, columnTypes(column))
+                        .getOrElse(throw new Exception(s"Unexpected value $value in column $column"))
+                  }
+                ProjectedSequenceRow(r.rowNumber, values)
+              }
+          }
+
+      fs2.Stream.eval(projectedRows).flatMap(r => r)
+    }
+
+  def fromRow(schema: StructType): (Array[String], Row => SequenceRow) = {
+    val rowNumberIndex = schema.fieldNames.indexOf("rowNumber")
+    if (rowNumberIndex < 0) {
+      throw new Exception("Can't upsert sequence rows, column `rowNumber` is missing.")
+    }
+
+    val columns = schema.fields.zipWithIndex.filter(_._1.name != "rowNumber").map {
+      case (field, index) =>
+        val seqColumn = sequenceInfo.columns
```

Comment:
Store `sequenceInfo.columns` as a `Set` and use `contains`.

Reply:
Turns out I already had the `columnTypes` map :)